<a href="https://colab.research.google.com/github/Raynould-Joseph/NLP/blob/main/RBM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  Step1: Train the network on the data of all the users.


#  Step2: Take the training data of a specific user during inference time.


#  Step3: Use the data to obtain the activations of the hidden neuron.


#  Step4: Use the hidden neuron values to get the activations of the input neurons.


#  Step5: The new values of input neurons show the rating the user would give.



# importing  torchvision



In [33]:
!pip install -q torch==1.0.0 torchvision

# Importing the libraries  

In [34]:

import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.nn.parallel
import torch.optim as optim
import torch.utils.data

In [35]:
!pip install torch

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [36]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [37]:
df = pd.read_csv('/content/gdrive/MyDrive/NLP lab/AmazonBooks - Sheet1csv.csv')
df

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
595  Live Free Or Die: America (and the World) on t...   
596                                 Burn After Writing   
597  Get Out of Your Head: Stopping the Spiral of T...   
598  Human Body Activity Book for Kids: Hands-On Fu...   
599                                 Educated: A Memoir   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
595              Sean Hannity          4.8    20153     15  2020  Non Fiction  
596              Sharon Jones          4.6    26423      8  2020  Non Fiction  
597              Jennie Allen          4.7    12879     14  2020  Non Fiction  
598              Katie Stokes          4.7    14128      5  2020  Non Fiction  
599            Tara Westover           4.7    45351     11  2020  Non Fiction  

[600 rows x 7 columns]

In [38]:
import numpy as np  
import pandas as pd  
import torch  
import torch.nn as nn  
import torch.nn.parallel  
import torch.optim as optim  
import torch.utils.data  
from torch.autograd import Variable  

# Importing the dataset 

import is all your movies, which are in the file movies.dat into a new variable called movies


then we will use the read_csv() function for reading the CSV file.

1.The first argument is the path that contains the dataset

2.The second argument is the separator, and the default separator is the comma that works for the CSV files where the features are separated by commas.

3.Then the third argument is the header because actually, the file movies.dat doesn't contain the header, i.e., names of columns. 

4.The next parameter is the engine, which is to make sure that the dataset gets imported correctly, so we will use the python engine to make it efficient.

5.UTF-8. So, we will input latin-1 due to some of the special characters in the movie's title.

In [39]:
# Importing the dataset  
movies = pd.read_csv('/content/gdrive/MyDrive/RBM/ml-1m/ml-1m/movies.dat', sep = '::', header = None, engine = 'python', encoding = 'latin-1') 
movies

0                                   1                             2
0        1                    Toy Story (1995)   Animation|Children's|Comedy
1        2                      Jumanji (1995)  Adventure|Children's|Fantasy
2        3             Grumpier Old Men (1995)                Comedy|Romance
3        4            Waiting to Exhale (1995)                  Comedy|Drama
4        5  Father of the Bride Part II (1995)                        Comedy
...    ...                                 ...                           ...
3878  3948             Meet the Parents (2000)                        Comedy
3879  3949          Requiem for a Dream (2000)                         Drama
3880  3950                    Tigerland (2000)                         Drama
3881  3951             Two Family House (2000)                         Drama
3882  3952               Contender, The (2000)                Drama|Thriller

[3883 rows x 3 columns]

Ratings followed by changing its path, and the rest will remain the same as it is.

In [40]:
Ratings = pd.read_csv('/content/gdrive/MyDrive/RBM/ml-1m/ml-1m//ratings.dat', sep = '::', header = None, engine = 'python', encoding = 'latin-1')  
Ratings

0     1  2          3
0           1  1193  5  978300760
1           1   661  3  978302109
2           1   914  3  978301968
3           1  3408  4  978300275
4           1  2355  5  978824291
...       ...   ... ..        ...
1000204  6040  1091  1  956716541
1000205  6040  1094  5  956704887
1000206  6040   562  5  956704746
1000207  6040  1096  4  956715648
1000208  6040  1097  4  956715569

[1000209 rows x 4 columns]

we will create a variable training_set followed by using the Pandas library to import u1.base. Then we will convert this training set into an array because by importing u1.base with Pandas, we will end up getting a DataFrame. 

So, first, we will use pandas read_csv function, and then we will pass our first argument, which is the path that will take the u1.base in the ml-100k folder and in order to do that, we will start with the folder that contains u1.base, which actually resides in the ml-100k folder followed by adding the name of the training set, i.e., u1.base.

delimiter = '\t' to specify the tab.

In [41]:
# Preparing the training set and the test set  
training_set = pd.read_csv('/content/gdrive/MyDrive/RBM/ml-100k/ml-100k/u1.base', delimiter = '\t')  
training_set

1   1.1  5  874965758
0        1     2  3  876893171
1        1     3  4  878542960
2        1     4  3  876893119
3        1     5  3  889751712
4        1     7  4  875071561
...    ...   ... ..        ...
79994  943  1067  2  875501756
79995  943  1074  4  888640250
79996  943  1188  3  888640250
79997  943  1228  3  888640275
79998  943  1330  3  888692465

[79999 rows x 4 columns]

Output:

1.the first column corresponds to the users

2.the second column corresponds to the movies

3. third column corresponds to the ratings

4.fourth column corresponds to the timestep
( we really don't need because it won't be relevant to train the model)

Next step: 

array to convert a DataFrame into an array

syntax

we will first input the training_set argument, and as a second argument, we will need to specify the type of this new array that we are creating.

In [42]:
training_set = np.array(training_set, dtype = 'int')  
training_set

array([[        1,         2,         3, 876893171],
       [        1,         3,         4, 878542960],
       [        1,         4,         3, 876893119],
       ...,
       [      943,      1188,         3, 888640250],
       [      943,      1228,         3, 888640275],
       [      943,      1330,         3, 888692465]])

# repeating the same process for the test data set

1. users in the first column.

2. movies in the second column

3.ratings in the third column. 

In [43]:
test_set = pd.read_csv('/content/gdrive/MyDrive/RBM/ml-100k/ml-100k/u1.test', delimiter = '\t')  


given below that our test_set is an array of integers32 of 20,000 ratings that correspond to the 20% of the original dataset composed of the 100,000 ratings.

In [44]:
test_set = np.array(test_set, dtype = 'int')  
test_set

array([[        1,        10,         3, 875693118],
       [        1,        12,         5, 878542960],
       [        1,        14,         5, 874965706],
       ...,
       [      459,       934,         3, 879563639],
       [      460,        10,         3, 882912371],
       [      462,       682,         5, 886365231]])

Output here:
we have to understand that both the test_set and the training_set have different ratings.


However, we have the same users. Here indeed, we start with user 1 as in the training_set, but for this same user 1, we won't have the same movies because the ratings are different.

we will be converting our training_set as well as the test_set into a matrix, where the lines are going to be users, the columns are going to be the movies and the cells are going to be the ratings. We are going to create such a matrix for the training_set and another one for the test_set.

In [45]:
# Getting the number of users and movies  
nb_users = int(max(max(training_set[:,0]), max(test_set[:,0])))  

Now, in the same we will do for the movies, we will use the same code but will replace the index of the column users, which is 0 by the index of the column movies, i.e., 1.

In [46]:
nb_movies = int(max(max(training_set[:,1]), max(test_set[:,1])))  

So, we had to take the max of the max because we don't know if this movie ID was in the training set or test set, and we actually check out by running the following command.

In [47]:
max(training_set[:,1])  


1682

Output:
Therefore, the maximum movie ID in the training_set is 1682.

In [48]:
max(test_set[:,1])  

1591

we will convert our training_set and test_set into an array with users in lines and movies in columns because we need to make a specific structure of data that will correspond to what the restricted Boltzmann machine expects as inputs. 

In [49]:
# converting the data into array with users in lines and movies in column
def convert(data):
 new_data = []
 for id_users in range(1, nb_users+1):
   id_movies = data[:,1][data[:,0] == id_users]
   id_ratings = data[:,2][data[:,0] == id_users]
   ratings = np.zeros(nb_movies)
   ratings[id_movies - 1] = id_ratings
   new_data.append(list(ratings))
   return new_data
 training_set = convert(training_set)
 test_set = convert(test_set)


output

1. we will create a structure that will contain these observations, which will go into the network, and their different features will go into the input nodes.

2.the observation in lines and the features in columns, which is exactly the structure of data expected by the neural network.

3.we will start with def, which stands for definition, followed by giving it a name called convert(). Inside the function, we will pass only one argument, i.e., data because we will apply this function to only a set, which will be the training_set first and then the test_set.



In [50]:
training_set

array([[        1,         2,         3, 876893171],
       [        1,         3,         4, 878542960],
       [        1,         4,         3, 876893119],
       ...,
       [      943,      1188,         3, 888640250],
       [      943,      1228,         3, 888640275],
       [      943,      1330,         3, 888692465]])

output:
training_set is a list of 943 lists.

From the above we can see this huge list contains 943 horizontal lists, where each of these 943 lists corresponds to each user of our database. So, we can check for the first movie, the second movie and the third movie; the ratings are as expected 0, 3 and 4. It can be clearly seen that for each user, we get the ratings of all the movies of the database, and we get a 0 when the movies weren't rated and the real rating when the user rated the movie.



In [51]:
# Converting the data into Torch tensors  
training_set = torch.FloatTensor(training_set)  

In [52]:
test_set = torch.FloatTensor(test_set)  


In [53]:
# Converting the ratings into binary ratings 1 (Liked) or 0 (Not Liked)  
training_set[training_set == 0] = -1  

In [54]:
training_set[training_set == 1] = 0  
training_set[training_set == 2] = 0  


In [55]:
training_set[training_set >= 3] = 1  

In [56]:
test_set[test_set == 0] = -1  
test_set[test_set == 1] = 0  
test_set[test_set == 2] = 0  
test_set[test_set >= 3] = 1  

In [57]:
# Creating the architecture of the Neural Network  
class RBM():  
    def __init__(self, nv, nh):  
      self.W = torch.randn(nh, nv)
      self.a = torch.randn(1, nh)    
      self.b = torch.randn(1, nv)  
    def sample_h(self, x):
      wx = torch.mm(x, self.W.t())   
      activation = wx + self.a.expand_as(wx)  
      p_h_given_v = torch.sigmoid(activation) 
      return p_h_given_v, torch.bernoulli(p_h_given_v)  
    def sample_v(self, y):
      wy = torch.mm(y, self.W)
      activation = wy + self.b.expand_as(wy)
      p_v_given_h = torch.sigmoid(activation)
      return p_v_given_h, torch.bernoulli(p_v_given_h)
    def train(self, v0, vk, ph0, phk):
      self.W += (torch.mm(v0.t(), ph0) - torch.mm(vk.t(), phk)).t()
      self.b += torch.sum((v0 - vk), 0)  
      self.a += torch.sum((ph0 - phk), 0) 
nv = len(training_set[0])
nh = 100
batch_size = 100
rbm = RBM(nv, nh)


In [58]:
# Training the RBM
nb_epoch = 10
for epoch in range(1, nb_epoch + 1):
    train_loss = 0
    s = 0.
    for id_user in range(0, nb_users - batch_size, batch_size):
        vk = training_set[id_user:id_user+batch_size]
        v0 = training_set[id_user:id_user+batch_size]
        ph0,_ = rbm.sample_h(v0)
        for k in range(10):
           _,hk = rbm.sample_h(vk)
           _,vk = rbm.sample_v(hk)
           vk[v0<0] = v0[v0<0]
        phk,_ = rbm.sample_h(vk)
        rbm.train(v0, vk, ph0, phk)
        train_loss += torch.mean(torch.abs(v0[v0>=0] - vk[v0>=0]))
        s += 1.
    print('epoch: '+str(epoch)+' loss: '+str(train_loss/s))

epoch: 1 loss: tensor(0.1214)
epoch: 2 loss: tensor(0.1228)
epoch: 3 loss: tensor(0.1161)
epoch: 4 loss: tensor(0.1139)
epoch: 5 loss: tensor(0.1228)
epoch: 6 loss: tensor(0.1161)
epoch: 7 loss: tensor(0.1367)
epoch: 8 loss: tensor(0.0950)
epoch: 9 loss: tensor(0.1161)
epoch: 10 loss: tensor(0.1128)


In [59]:
# Testing the RBM  
test_loss = 0  
s = 0.  

In [60]:
for id_user in range(nb_users): 
  v = training_set[id_user:id_user+1]  
  vt = test_set[id_user:id_user+1]
  if len(vt[vt>=0]) > 0:  
    _,h = rbm.sample_h(v)  
    _,v = rbm.sample_v(h)  
    test_loss += torch.mean(torch.abs(vt[vt>=0] - v[vt>=0]))  
    s += 1.  
print('test loss: '+str(test_loss/s))  

test loss: tensor(0.2505)
